File: src/PrinciPal.McpServer/Services/SessionManager.cs
using System.Collections.Concurrent;
using PrinciPal.Contracts;

namespace PrinciPal.McpServer.Services;

/// <summary>
/// Manages multiple VS debug sessions. Each session is keyed by a unique ID
/// (short hash of the solution path) and has a friendly name (solution filename).
/// Each session gets its own <see cref="DebugStateStore"/> instance.
/// </summary>
public class SessionManager
{
    private readonly ConcurrentDictionary<string, SessionEntry> _sessions = new(StringComparer.OrdinalIgnoreCase);

    public int SessionCount => _sessions.Count;

    /// <summary>
    /// Gets or creates a session, returning its <see cref="DebugStateStore"/>.
    /// Auto-registers the session on first access.
    /// </summary>
    public DebugStateStore GetOrCreateSession(string sessionId, string? name = null, string? solutionPath = null)
    {
        var entry = _sessions.GetOrAdd(sessionId, id => new SessionEntry
        {
            Store = new DebugStateStore(),
            Info = new SessionInfo
            {
                SessionId = id,
                Name = name ?? "",
                SolutionPath = solutionPath ?? "",
                ConnectedAt = DateTime.UtcNow
            }
        });

        // Update metadata if provided and was previously empty
        if (!string.IsNullOrEmpty(name) && string.IsNullOrEmpty(entry.Info.Name))
            entry.Info.Name = name;
        if (!string.IsNullOrEmpty(solutionPath) && string.IsNullOrEmpty(entry.Info.SolutionPath))
            entry.Info.SolutionPath = solutionPath;

        return entry.Store;
    }

    /// <summary>
    /// Gets a session's store by its unique ID, or null if not found.
    /// </summary>
    public DebugStateStore? GetSession(string sessionId)
    {
        return _sessions.TryGetValue(sessionId, out var entry) ? entry.Store : null;
    }

    /// <summary>
    /// Resolves a query string (name or ID) to a session store.
    /// Returns (store, null) on success, or (null, errorMessage) on failure.
    /// </summary>
    public (DebugStateStore? Store, string? Error) ResolveByNameOrId(string query)
    {
        // Try exact ID match first
        if (_sessions.TryGetValue(query, out var entry))
            return (entry.Store, null);

        // Try name match
        var matches = _sessions.Values
            .Where(e => string.Equals(e.Info.Name, query, StringComparison.OrdinalIgnoreCase))
            .ToList();

        if (matches.Count == 1)
            return (matches[0].Store, null);

        if (matches.Count > 1)
        {
            var lines = matches.Select(m => $"  {m.Info.Name} [{m.Info.SessionId}] - {m.Info.SolutionPath}");
            return (null, $"Multiple sessions named '{query}'. Use the session ID instead:\n{string.Join("\n", lines)}");
        }

        return (null, $"Session '{query}' not found. Use list_sessions to see connected VS instances.");
    }

    /// <summary>
    /// Removes a session and all its state.
    /// </summary>
    public bool RemoveSession(string sessionId)
    {
        return _sessions.TryRemove(sessionId, out _);
    }

    /// <summary>
    /// Returns info for all active sessions. HasDebugState is computed dynamically.
    /// </summary>
    public List<SessionInfo> GetAllSessions()
    {
        var result = new List<SessionInfo>();
        foreach (var kvp in _sessions)
        {
            var info = kvp.Value.Info;
            var state = kvp.Value.Store.GetCurrentState();
            result.Add(new SessionInfo
            {
                SessionId = info.SessionId,
                Name = info.Name,
                SolutionPath = info.SolutionPath,
                ConnectedAt = info.ConnectedAt,
                HasDebugState = state is { IsInBreakMode: true }
            });
        }
        return result;
    }

    private class SessionEntry
    {
        public required DebugStateStore Store { get; init; }
        public required SessionInfo Info { get; init; }
    }
}

File: mcp_server/src/session_manager.ts
import { debug_state_store } from "./debug_state_store.js";
import type { session_info } from "./domain_types.js";

type session_entry = {
    store: debug_state_store;
    info: {
        session_id: string;
        name: string;
        solution_path: string;
        connected_at: Date;
        last_seen: Date;
    };
};

/**
 * `mode` "open": agent may start any name in `available` except those in `blocked`.
 * `mode` "strict": agent may start only names in `allowed`.
 */
type launch_control_state = {
    available: string[];
    mode: "open" | "strict";
    blocked: string[];
    allowed: string[];
};

export type resolve_store_result =
    | { ok: true; store: debug_state_store }
    | { ok: false; description: string };

/**
 * Multi-session registry (port of C# SessionManager).
 */
export class session_manager {
    private readonly sessions = new Map<string, session_entry>();
    private readonly launch_by_session = new Map<string, launch_control_state>();

    get session_count(): number {
        return this.sessions.size;
    }

    get_or_create_session(
        session_id: string,
        name?: string | null,
        solution_path?: string | null
    ): debug_state_store {
        const now = new Date();
        let entry = this.sessions.get(session_id);
        if (!entry) {
            entry = {
                store: new debug_state_store(),
                info: {
                    session_id,
                    name: name ?? "",
                    solution_path: solution_path ?? "",
                    connected_at: now,
                    last_seen: now,
                },
            };
            this.sessions.set(session_id, entry);
        } else {
            if (name && !entry.info.name) entry.info.name = name;
            if (solution_path && !entry.info.solution_path) entry.info.solution_path = solution_path;
            entry.info.last_seen = new Date();
        }
        return entry.store;
    }

    get_session(session_id: string): debug_state_store | undefined {
        return this.sessions.get(session_id)?.store;
    }

    remove_session(session_id: string): void {
        this.sessions.delete(session_id);
        this.launch_by_session.delete(session_id);
    }

    resolve_by_name_or_id(query: string): resolve_store_result {
        const exact = this.sessions.get(query);
        if (exact) return { ok: true, store: exact.store };

        const matches = [...this.sessions.values()].filter((e) =>
            e.info.name.toLowerCase() === query.toLowerCase()
        );
        if (matches.length === 1) return { ok: true, store: matches[0].store };
        if (matches.length > 1) {
            const lines = matches.map(
                (m) => `  ${m.info.name} [${m.info.session_id}] - ${m.info.solution_path}`
            );
            return {
                ok: false,
                description: `Session "${query}" is ambiguous. Matches:\n${lines.join("\n")}`,
            };
        }
        return { ok: false, description: `Session not found: ${query}` };
    }

    /** Workspace session id (8-char hash) for agent commands / long-poll. */
    resolve_session_id(query: string): string | undefined {
        const r = this.resolve_by_name_or_id(query);
        if (!r.ok) return undefined;
        for (const [id, e] of this.sessions) {
            if (e.store === r.store) return id;
        }
        return undefined;
    }

    get_all_sessions(): session_info[] {
        const result: session_info[] = [];
        for (const entry of this.sessions.values()) {
            const state = entry.store.get_current_state();
            result.push({
                session_id: entry.info.session_id,
                name: entry.info.name,
                solution_path: entry.info.solution_path,
                connected_at: entry.info.connected_at.toISOString(),
                last_seen: entry.info.last_seen.toISOString(),
                has_debug_state: !!state?.is_in_break_mode,
            });
        }
        return result;
    }

    get_stale_session_ids(timeout_ms: number): string[] {
        const cutoff = Date.now() - timeout_ms;
        return [...this.sessions.entries()]
            .filter(([, e]) => e.info.last_seen.getTime() < cutoff)
            .map(([id]) => id);
    }

    set_launch_sync(session_id: string, available: string[]): void {
        const prev = this.launch_by_session.get(session_id);
        this.launch_by_session.set(session_id, {
            available: [...available],
            mode: prev?.mode ?? "open",
            blocked: prev?.blocked ?? [],
            allowed: prev?.allowed ?? [],
        });
    }

    set_launch_policy(
        session_id: string,
        mode: "open" | "strict",
        blocked: string[],
        allowed: string[]
    ): void {
        const prev = this.launch_by_session.get(session_id);
        this.launch_by_session.set(session_id, {
            available: prev?.available ?? [],
            mode,
            blocked: [...blocked],
            allowed: [...allowed],
        });
    }

    get_launch_control(session_id: string): launch_control_state | undefined {
        return this.launch_by_session.get(session_id);
    }

    is_launch_allowed_for_agent(session_id: string, name: string): boolean {
        const st = this.launch_by_session.get(session_id);
        if (!st || st.available.length === 0) return false;
        if (!st.available.includes(name)) return false;
        if (st.mode === "open") return !st.blocked.includes(name);
        return st.allowed.includes(name);
    }
}
